collections/index.js

```diff
@@ -20,46 +20,130 @@ import employees from './data.json';
   of employees who are missing any data
 */
-export const getEmployeesWithMissingData = () => {
 
+export const getEmployeesWithMissingData1 = () => {
+  // sets an empty array ready to accept any values we add later:
+  var missingInfoEmployees = [];
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // checks if any of the values are empty. I prefer this method over others
+    // I thought of because if more than one value is empty per employee, this
+    // function only adds the employee to the array once.
+    if (employees[employeeNumberIndex].id === "" || employees[employeeNumberIndex].firstName === "" || employees[employeeNumberIndex].lastName === "" || employees[employeeNumberIndex].department === "" || employees[employeeNumberIndex].contactNumber === "" || employees[employeeNumberIndex].taxCode === "" || employees[employeeNumberIndex].salary === "") {
+      // this adds (pushes) the employees details to the missingInfoEmployees
+      // array we created earlier.
+      missingInfoEmployees.push(employees[employeeNumberIndex]);
+    }
+  }
+  // returns all employees that are missing data
+  return missingInfoEmployees;
+}
+
+// I thought of a 2nd way of doing this which is probably better but might
+// need some tidying up.
+
+export const getEmployeesWithMissingData = () => {
+  // sets an empty array ready to accept any values we add later:
+  var missingInfoEmployees = [];
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // creates an array of the employee details
+    let employeeDetails = Object.values(employees[employeeNumberIndex])
+    // checks each value in the new array for any empty strings
+    employeeDetails.find(function(item) {
+      if (item === "") {
+        // adds the employee to the missingInfoEmployees array
+        missingInfoEmployees.push(employees[employeeNumberIndex]);
+      }
+    })
+  }
+  // returns all employees that are missing data
+  return missingInfoEmployees;
 }
 
 /*
   I expect this functon to return me the employee list but with any sensitive
-  information like personal phone numbers or salary inforamtion stripped out
+  information like personal phone numbers or salary and tax inforamtion
+  stripped out
 */
 export const getNonSensitiveInformation = () => {
-
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // deletes the 3 sensitive key/value pairs from the employee records
+    delete employees[employeeNumberIndex].contactNumber;
+    delete employees[employeeNumberIndex].taxCode;
+    delete employees[employeeNumberIndex].salary;
```

Comment:
You're actually modifying the original data set here which is generally not what we want to be doing - if we ever wanna take a set of data and copy it but modify each element we wanna be using `array.map` to keep the integrity of the original data intact

Reply:
What you will find when doing array.map also is that rather than getting a set of data and **deleting what we don't want** - we're creating a **new set of data of what we do want**.

If you imagine using this problem in a real world situation where we want to hide sensitive information from non-managers, if we delete what we don't want then every time we add a new piece of information to a user we have to go back and retroactively apply that to this functon to make sure a user doesn't see it which can maybe get missed and users might end up seeing data they shouldn't because this was forgotten. However if we did it the other way and built up a new object of what we want to see, the only way a user would see it is if we want back and added it in - therefore meaning there's less room for error